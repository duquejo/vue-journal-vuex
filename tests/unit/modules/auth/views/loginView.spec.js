import { shallowMount } from '@vue/test-utils';
import LoginView from '@/modules/auth/views/LoginView';
import createVuexStore from '../../../mocks/mock-store';
import { useRouter } from 'vue-router';

import Swal from 'sweetalert2';

jest.mock('vue-router', () => ({
  useRoute: jest.fn(),
  useRouter: jest.fn(() => ({
    push: () => {}
  }))
}));

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));

describe('LoginView.vue', () => {

  const store = createVuexStore({
    status: 'not-authenticated',
    user: null,
    idToken: null,
    refreshToken: null,
  });

  store.dispatch = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render the component', () => {
    const wrapper = shallowMount(LoginView, {
      global: {
        plugins: [ store ],
        stubs: ['router-link']
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should display an error if login fails', async () => {

    const push = jest.fn();
    useRouter.mockImplementationOnce(() => ({ push }));
    store.dispatch.mockReturnValueOnce({ ok: false, message: 'EMAIL/PASSWORD Credentials'});
    const wrapper = shallowMount(LoginView, {
      global: {
        plugins: [ store ],
        stubs: ['router-link']
      }
    });

    await wrapper.find('form').trigger('submit');

    /**
     * Taking advantage of the mocked Storage.
     */
    expect(store.dispatch).toHaveBeenCalledWith('auth/signInUser', {
      email: '',
      password: '',
    });
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'EMAIL/PASSWORD Credentials', 'error');
    expect(push).not.toHaveBeenCalled();
  });

  test('should signIn an user and redirect to the main journal view', async () => {

    const push = jest.fn();
    useRouter.mockImplementationOnce(() => ({ push }));
    store.dispatch.mockReturnValueOnce({ ok: true });
    const wrapper = shallowMount(LoginView, {
      global: {
        plugins: [ store ],
        stubs: ['router-link']
      }
    });

    const [ inputEmail, inputPassword ] = wrapper.findAll('input');
    await inputEmail.setValue('foo@bar.com');
    await inputPassword.setValue('foo-123');

    await wrapper.find('form').trigger('submit');

    /**
     * Taking advantage of the mocked Storage.
     */
    expect(store.dispatch).toHaveBeenCalledWith('auth/signInUser', {
      email: 'foo@bar.com',
      password: 'foo-123',
    });
    expect(Swal.fire).not.toHaveBeenCalled();
    expect(push).toHaveBeenCalledWith({ name: 'no-entry' });

    expect(wrapper.vm.userForm).toHaveProperty('email', 'foo@bar.com');
    expect(wrapper.vm.userForm).toHaveProperty('password', 'foo-123');
  });
});