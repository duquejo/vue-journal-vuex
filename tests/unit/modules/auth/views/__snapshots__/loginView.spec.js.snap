// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView.vue should render the component 1`] = `
<span class="login100-form-title p-b-41">Daybook access</span>
<form class="login100-form validate-form p-b-33 p-t-5" autocomplete="off">
  <div class="wrap-input100 validate-input" data-validate="Enter password"><input class="input100" type="text" placeholder="Email" required=""><span class="focus-input100" data-placeholder=""></span></div>
  <div class="wrap-input100 validate-input" data-validate="Enter password"><input class="input100" type="password" placeholder="Password" required=""><span class="focus-input100" data-placeholder=""></span></div>
  <div class="container-login100-form-btn m-t-32"><button class="login100-form-btn"> Login </button></div>
  <div class="container-login100-form-btn m-t-32">
    <router-link-stub to="[object Object]"></router-link-stub>
  </div>
</form>
`;
